### PR TITLE
Week 4

### PR DESCRIPTION
**Pull request by sl2902**
- Add cli for SQLModelRepository
- Add unit tests for cli
- Add backend script to choose between repositories
- Add unit tests for backend

### CODE REVIEW

File: src/snipster/cli.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        Snippet(
+            title=title,
+            code=code,
+            description=description,
+            language=language,
+            tags=tags,
+            favorite=favorite,
+        )
+    )
+    # typer.secho(f" Added snippet: {title}")
```
Comment (juanjoseexpositogonzalez) [inline src/snipster/cli.py]: **Class: GOOD TO HAVE**

**What to review:**
```python
# typer.secho(f" Added snippet: {title}")
```

**Explanation:**
Commented-out code should be removed from the codebase. It adds noise and can become outdated. If the code is needed for reference, it should be in version control history.

**Fix proposal:**
Remove the commented line:
```python
repo.add(...)
console.print(
    f":white_check_mark: [bold green]Snippet '{title}' added[/bold green]"
)
```

File: src/snipster/cli.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+def tags(
+    snippet_id: int = typer.Option(..., prompt=True),
+    tags: str = typer.Option(..., prompt=True, help="Add comma separated tags"),
+    remove: bool = typer.Option(
+        False, prompt=True, help="True to remove tags. Default is False"
+    ),
+    sort: bool = typer.Option(True, help="True if tags are sorted. Default is True"),
+    ctx: typer.Context = None,
+):
+    """Add comma separted tags to snippet"""
```
Comment (juanjoseexpositogonzalez) [inline src/snipster/cli.py]: **Class: GOOD TO HAVE**

**What to review:**
```python
"""Add comma separted tags to snippet"""
```

**Explanation:**
Typo in docstring: "separted" should be "separated".

**Fix proposal:**
```python
"""Add comma separated tags to snippet"""
```

File: src/snipster/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+"""Custom exceptions"""
+
+
+class SnippetNotFoundError(Exception):
```
Comment (juanjoseexpositogonzalez) [inline src/snipster/exceptions.py]: **Class: GOOD TO HAVE**

**What to review:**
```python
class SnippetNotFoundError(Exception):
    pass
```

**Explanation:**
The custom exception lacks a docstring explaining when it should be raised. Adding a docstring improves code documentation and helps other developers understand the exception's purpose.

**Fix proposal:**
```python
class SnippetNotFoundError(Exception):
    """
    Raised when a snippet with the specified ID does not exist in the repository.

    This exception is used by repository implementations to signal that a
    requested snippet could not be found.
    """
    pass
```

File: tests/test_cli.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    )
+
+    mock_repo.delete.return_value = snippet
+    result = runner.invoke(app, ["delete", "--snippet-id", 1])
+
+    assert result.exit_code == 0
+    assert "deleted" in result.stdout
+
+
+def test_delete_snippet_not_found(mock_repo):
```
Comment (juanjoseexpositogonzalez) [inline tests/test_cli.py]: **Class: NEEDED**

**What to review:**
```python
def test_delete_snippet_not_found(mock_repo):
    """Test deleting non-existent snippet"""
    from sqlalchemy.exc import NoResultFound

    mock_repo.delete.side_effect = NoResultFound()

    result = runner.invoke(app, ["delete", "--snippet-id", 1])
    assert result.exit_code == 1
    assert "not found" in result.stdout
```

**Explanation:**
This test uses `NoResultFound` exception, but the actual code path raises `KeyError` from `database_manager.delete_record`. The test is testing behavior that doesn't match the actual implementation (see related CLI review item about exception handling).

**Fix proposal:**
Either update the test to use `KeyError` to match actual behavior, or update the CLI to catch the appropriate exception:

```python
def test_delete_snippet_not_found(mock_repo):
    """Test deleting non-existent snippet"""
    mock_repo.delete.side_effect = KeyError("Record not found")

    result = runner.invoke(app, ["delete", "--snippet-id", 1])
    assert result.exit_code == 1
    assert "not found" in result.stdout
```

File: tests/test_cli.py
Comment (juanjoseexpositogonzalez) [inline tests/test_cli.py]: **Class: GOOD TO HAVE**

**What to review:**
Missing tests for edge cases.

**Explanation:**
Some command tests don't verify the full behavior. Consider adding:
- Test for `add` command with invalid language
- Test for `add` command with all optional parameters
- Test for `search` with language filter
- Test for `list` with verbose flag enabled

**Fix proposal:**
Add additional test cases:
```python
def test_add_with_all_options(mock_repo):
    """Test adding a snippet with all options"""
    result = runner.invoke(
        app,
        [
            "add",
            "--title", "Test",
            "--code", "print('hello')",
            "--description", "A test snippet",
            "--language", "Python",
            "--tags", "test,example",
            "--favorite",
        ],
    )

    assert result.exit_code == 0
    mock_repo.add.assert_called_once()
    call_args = mock_repo.add.call_args[0][0]
    assert call_args.favorite is True
    assert call_args.tags == "test,example"


def test_search_with_language_filter(mock_repo):
    """Test searching with language filter"""
    snippet = Snippet(id=1, title="Test", code="code", language=Language.PYTHON)
    mock_repo.search.return_value = [snippet]

    result = runner.invoke(
        app, ["search", "--term", "test", "--language", "python"]
    )

    assert result.exit_code == 0
    mock_repo.search.assert_called_once_with("test", language="python")
```

**Review by juanjoseexpositogonzalez — COMMENTED** [review]
Great work on implementing the CLI and backend factory pattern! The code structure is clean and the test coverage is good. Here are the main findings:

### Critical Issues (Must Fix)

1. **Exception Mismatch in CLI delete command** (`cli.py:138`)
   - The CLI catches `NoResultFound` but `database_manager.delete_record` raises `KeyError`
   - This means delete errors won't be handled properly

2. **Missing f-string prefix** (`sql_model_repository.py:35`)
   - `logger.info("Record id {id} deleted successfully")` logs literal `{id}` instead of the actual ID

### Required Changes

- [ ] Fix exception handling in CLI delete command (catch `KeyError` or use `SnippetNotFoundError`)
- [ ] Add `f` prefix to log statement in `sql_model_repository.py:35`
- [ ] Fix typo in class name: `ReposistoryType` → `RepositoryType` (`backend.py:13`)
- [ ] Remove SQLAlchemy imports from CLI to maintain abstraction layer
- [ ] Fix variable shadowing: `list` function name (`cli.py:78`) and `tags` variable (`cli.py:213`)
- [ ] Fix double assignment in tests (`test_cli.py:57, 80-81`)

### Suggestions (Good to Have)

- Toggle favourite message should reflect actual state change ("favourited" vs "unfavourited")
- Use modern type hints (`list[T]` instead of `List[T]`)
- Add docstrings to custom exceptions
- Simplify toggle logic: `snippet.favorite = not snippet.favorite`

See individual file reviews for detailed explanations and fix proposals.

---

## Issues by Priority

### CRITICAL (2 issues)

| File | Line | Issue | Impact |
|------|------|-------|--------|
| `cli.py` | 138 | Catches wrong exception type (`NoResultFound` vs `KeyError`) | Delete errors show as unhandled exceptions |
| `sql_model_repository.py` | 35 | Missing f-string prefix in log | Debugging issues, logs show `{id}` literally |

### NEEDED (8 issues)

| File | Line | Issue |
|------|------|-------|
| `cli.py` | 7 | SQLAlchemy imports break abstraction |
| `cli.py` | 78 | Function `list` shadows built-in |
| `cli.py` | 190 | Misleading toggle message |
| `cli.py` | 213 | Variable `tags` shadowed |
| `backend.py` | 13 | Typo: `ReposistoryType` |
| `database_manager.py` | 283 | Missing f-string prefix |
| `test_cli.py` | 57, 80-81 | Double assignment copy-paste error |
| `test_backend.py` | 41-47 | Creates real DB file in tests |

### GOOD TO HAVE (15+ issues)

See individual file reviews for complete list.

---

## Architecture Observations

### Positives

1. **Good use of Factory Pattern** - The `create_repository` function cleanly abstracts repository creation
2. **Consistent Repository Interface** - All repositories implement `SnippetRepository` ABC
3. **DRY Principle Applied** - `process_tags` moved to base class to avoid duplication
4. **Good Test Structure** - Tests are well-organized with fixtures and clear naming

### Areas for Improvement

1. **Leaky Abstraction** - CLI imports SQLAlchemy exceptions directly, coupling it to the database layer
2. **Inconsistent Exception Handling** - Some methods raise `KeyError`, others raise `SnippetNotFoundError`
3. **Missing Validation** - `update` method doesn't validate column existence like `select_with_filter` does

---

## Recommended Fix Order

1. **First** - Fix the CRITICAL issues (exception mismatch, missing f-string)
2. **Second** - Fix the typo in `RepositoryType` class name (public API)
3. **Third** - Address abstraction layer issues (remove SQLAlchemy from CLI)
4. **Fourth** - Clean up code quality issues (shadowing, double assignments)
5. **Finally** - Apply GOOD TO HAVE improvements

---

## Files Changed Summary

| File | Additions | Type | Review Status |
|------|-----------|------|---------------|
| `src/snipster/cli.py` | 226 | New | Reviewed - Issues found |
| `src/snipster/repositories/backend.py` | 46 | New | Reviewed - Issues found |
| `src/snipster/exceptions.py` | 5 | New | Reviewed - Minor suggestions |
| `src/snipster/database_manager.py` | 22 | Modified | Reviewed - Issues found |
| `src/snipster/repositories/sql_model_repository.py` | 17 | Modified | Reviewed - Issues found |
| `src/snipster/repositories/repository.py` | 23 | Modified | Reviewed - Minor suggestions |
| `src/snipster/repositories/in_memory_repository.py` | 1 | Modified | OK |
| `tests/test_cli.py` | 203 | New | Reviewed - Issues found |
| `tests/test_backend.py` | 53 | New | Reviewed - Issues found |
| `tests/repositories/test_sql_model_repository.py` | 51 | Modified | OK |
| `tests/test_database_manager.py` | 29 | Modified | OK |

---

## Additional Notes

### Binary Files Committed

The PR includes two database files that should probably be in `.gitignore`:
- `snippets.db`
- `test_from_env.db`

**Recommendation:** Add `*.db` to `.gitignore` and remove these files from the repository.

### Test Data File

`data/all_snippets.jsonl` contains test data. Consider whether this should be:
- In a `tests/fixtures/` directory
- Generated dynamically in tests
- Kept as-is for manual testing purposes

File: src/snipster/repositories/sql_model_repository.py
[the remark this answers is not in the document]
Comment (sl2902) [reply src/snipster/repositories/sql_model_repository.py]: For the last suggestion, given that there is a base abstract function called list results in a conflict. So I have to use List[Snippet]

File: tests/test_cli.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    mock_repo.list.return_value = []
+
+    result = runner.invoke(app, ["list"])
+    assert result.exit_code == 0
+    assert "No snippets found" in result.stdout
+
+
+def test_get_snippet_found(mock_repo):
+    """Test getting a snippet by ID"""
+    snippet = snippet = Snippet(
```
[the remark this answers is not in the document]
Comment (sl2902) [reply tests/test_cli.py]: nice catch!

**Review by juanjoseexpositogonzalez — COMMENTED** [review]
Reviewed

File: src/snipster/cli.py
[the remark this answers is not in the document]
Comment (juanjoseexpositogonzalez) [reply src/snipster/cli.py]: And was it working before the proposed change?

File: src/snipster/repositories/sql_model_repository.py
[the remark this answers is not in the document]
Comment (juanjoseexpositogonzalez) [reply src/snipster/repositories/sql_model_repository.py]: Yes, but there's a workaround in that (I sent the suggestion too) so the `list` command does not interfere with the `list` keyword in Python.

[the remark this answers is not in the document]
Comment (sl2902) [reply src/snipster/repositories/sql_model_repository.py]: I missed the suggestion. I don't see it here. Is it to use the following `from __future__ import annotations`?
